extensions/ql-vscode/test/vscode-tests/minimal-workspace/debugger/save-before-start.test.ts

```diff
@@ -0,0 +1,137 @@
+import * as path from "path";
+import {
+  Position,
+  TextDocument,
+  ViewColumn,
+  commands,
+  window,
+  workspace,
+} from "vscode";
+import { saveAllInGroup } from "../../../../src/run-queries-shared";
+
+/**
+ * Returns a new `TextDocument` with the given file name.
+ *
+ * @param file Path to the file to open, or undefined to create an untitled document.
+ * @param isDirty True if the document should have an edit applied.
+ * @param viewColumn The `ViewColumn` in which to open the file's editor.
+ * @returns The new `TextDocument`
+ */
+async function mockDocument(
+  file: string | undefined,
+  isDirty: boolean,
+  viewColumn: ViewColumn,
+): Promise<TextDocument> {
+  let doc: TextDocument;
+  if (file !== undefined) {
+    doc = await workspace.openTextDocument(
+      path.join(workspace.workspaceFolders![0].uri.fsPath, file),
+    );
+    const editor = await window.showTextDocument(doc, viewColumn, true);
+    if (isDirty) {
+      await editor.edit((edit) => {
+        edit.insert(new Position(0, 0), "test");
+      });
+    }
+  } else {
+    doc = await workspace.openTextDocument({
+      content: 'select "untitled"',
+      language: "ql",
+    });
+    await window.showTextDocument(doc, viewColumn, true);
+  }
+
+  return doc;
+}
+
+/**
+ * Returns a promise that resolves after the given number of milliseconds, and returns the string
+ * "timeout".
+ */
+function delay(ms: number): Promise<string> {
+  return new Promise((resolve) => setTimeout(() => resolve("timeout"), ms));
+}
+
+jest.setTimeout(10000); // A little extra time for the save dialog to appear.
+
+const dirtyFile = "debugger/dirty.ql";
+const cleanFile = "debugger/clean.ql";
+const otherGroupDirtyFile = "debugger/other-dirty.ql";
+const otherGroupCleanFile = "debugger/other-clean.ql";
+
+describe("saveBeforeStart", () => {
+  // We can't easily mock `TextDocument` because its properties are read-only and/or not
+  // configurable, so we rely on the actual `save()` method and the actual `isDirty` property.
+
+  beforeEach(async () => {
+    await commands.executeCommand("workbench.action.closeAllEditors");
+  });
+
+  it("should not save untitled documents without `includeUntitled`", async () => {
+    const dirtyDoc = await mockDocument(dirtyFile, true, ViewColumn.One);
+    const cleanDoc = await mockDocument(cleanFile, false, ViewColumn.One);
```

Comment:
Should this also include a dirty, untitled doc?

Reply:
This file has been removed.